### PR TITLE
Make softfailed a proper final result

### PR DESCRIPTION
**Pull request by coolo**
So far we had plenty of places that calculated a different result for
those.

### CODE REVIEW

**Comment by okurz** [comment]
LGTM

**Comment by coolo** [comment]
what I don't know is what will happen if the dashboard finds a softfailed result :)

File: dbicdh/_common/upgrade/40-41/001-set-softfails.pl [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+use strict;
+use OpenQA::Schema;
+use DBIx::Class::DeploymentHandler;
+use OpenQA::Schema::Result::Jobs;
+
+sub {
+    my $schema = shift;
+
+    # resort to raw SQL to avoid the schema freaking out
+    my $dents = $schema->resultset('JobModules')->search([' soft_failure != 0 ']);
```
Comment (aaannz) [inline dbicdh/_common/upgrade/40-41/001-set-softfails.pl]: Is this valid? Shouldn't be like `{  soft_failure => { '!=', 0}}` ?

**Comment by coolo** [comment]
ok, it will keep testing :)

```
notdone += 1 unless %w(passed failed).include?(job.result)

File: dbicdh/_common/upgrade/40-41/001-set-softfails.pl (hunk 1 above)
Comment (coolo) [inline dbicdh/_common/upgrade/40-41/001-set-softfails.pl]: see the comment?

Comment (coolo) [inline dbicdh/_common/upgrade/40-41/001-set-softfails.pl]: i.e. the syntax you propose doesn't work in migrations as it's using HEAD's schema

Comment (coolo) [inline dbicdh/_common/upgrade/40-41/001-set-softfails.pl]: (which is one of the reasons I said in the call that I don't expect too many members of the team to write proper migrations ;(

**Comment by okurz** [comment]
will this *automagically* enable carry over for soft fails, too?

**Comment by coolo** [comment]
no. carry over explicitly only cares for FAILED - and it's a rather tricky question what a carry over candidate is for soft fails. We don't even show the soft failed module in overview

**Comment by okurz** [comment]
well, that's fine. I would much more wish for a carry-over for non-fatal failed modules (see newly created progress issue)

**Comment by aaannz** [comment]
I tried this, but the migration failed:

> Uncaught exception from user code:
        failed to run Perl in /usr/share/openqa/script/../dbicdh/_common/upgrade/40-41/001-set-softfails.pl: DBIx::Class::Storage::DBI::_dbh_execute(): DBI Exception: DBD::Pg::st execute failed: ERROR:  operator does not exist: integer <> boolean
        LINE 1: ...b_modules SET result = $1 WHERE ( (  soft_failure != 0  IS N...
                                                                     ^
        HINT:  No operator matches the given name and argument type(s). You might need to add explicit type casts. [for Statement "UPDATE job_modules SET result = ? WHERE ( (  soft_failure != 0  IS NULL AND result = ? ) )" with ParamValues: 1='softfailed', 2='passed'] at /usr/share/openqa/script/../dbicdh/_common/upgrade/40-41/001-set-softfails.pl line 29
        DBIx::Class::DeploymentHandler::DeployMethod::SQL::Translator::catch {...} ('DBIx::Class::Exception=HASH(0x89c7a58)') called at /usr/lib/perl5/vendor_perl/5.18.2/Try/Tiny.pm line 104
        Try::Tiny::try('CODE(0x87a02c8)', 'Try::Tiny::Catch=REF(0x888be20)') called at /usr/lib/perl5/vendor_perl/5.18.2/DBIx/Class/DeploymentHandler/DeployMethod/SQL/Translator.pm line 322
        DBIx::Class::DeploymentHandler::DeployMethod::SQL::Translator::_run_perl('DBIx::Class::DeploymentHandler::DeployMethod::SQL::Translator...', 'Path::Class::File=HASH(0x89907f0)', 'ARRAY(0x4b5c0a8)') called at /usr/lib/perl5/vendor_perl/5.18.2/DBIx/Class/DeploymentHandler/DeployMethod/SQL/Translator.pm line 343
        DBIx::Class::DeploymentHandler::DeployMethod::SQL::Translator::__ANON__() called at /usr/lib/perl5/vendor_perl/5.18.2/DBIx/Class/DeploymentHandler/DeployMethod/SQL/Translator.pm line 331
        DBIx::Class::DeploymentHandler::DeployMethod::SQL::Translator::__ANON__() called at /usr/lib/perl5/vendor_perl/5.18.2/Context/Preserve.pm line 42
        Context::Preserve::preserve_context('CODE(0x8990070)', 'after', 'CODE(0x8724a00)') called at /usr/lib/perl5/vendor_perl/5.18.2/DBIx/Class/DeploymentHandler/DeployMethod/SQL/Translator.pm line 331
        DBIx::Class::DeploymentHandler::DeployMethod::SQL::Translator::txn_do('DBIx::Class::DeploymentHandler::DeployMethod::SQL::Translator...', 'CODE(0x8852cc0)') called at /usr/lib/perl5/vendor_perl/5.18.2/DBIx/Class/DeploymentHandler/DeployMethod/SQL/Translator.pm line 355
        DBIx::Class::DeploymentHandler::DeployMethod::SQL::Translator::_run_sql_and_perl('DBIx::Class::DeploymentHandler::DeployMethod::SQL::Translator...', 'ARRAY(0x3e76448)', undef, 'ARRAY(0x4b5c0a8)') called at /usr/lib/perl5/vendor_perl/5.18.2/DBIx/Class/DeploymentHandler/DeployMethod/SQL/Translator.pm line 699
        DBIx::Class::DeploymentHandler::DeployMethod::SQL::Translator::upgrade_single_step('DBIx::Class::DeploymentHandler::DeployMethod::SQL::Translator...', 'HASH(0x898ea80)') called at /usr/lib/perl5/vendor_perl/5.18.2/x86_64-linux-thread-multi/Moose/Meta/Method/Delegation.pm line 110
        DBIx::Class::DeploymentHandler::upgrade_single_step('DBIx::Class::DeploymentHandler=HASH(0x84b1470)', 'HASH(0x898ea80)') called at /usr/lib/perl5/vendor_perl/5.18.2/DBIx/Class/DeploymentHandler/Dad.pm line 64
        DBIx::Class::DeploymentHandler::Dad::__ANON__() called at /usr/lib/perl5/vendor_perl/5.18.2/DBIx/Class/DeploymentHandler/DeployMethod/SQL/Translator.pm line 331
        DBIx::Class::DeploymentHandler::DeployMethod::SQL::Translator::__ANON__() called at /usr/lib/perl5/vendor_perl/5.18.2/Context/Preserve.pm line 22
        Context::Preserve::preserve_context('CODE(0x8853128)', 'after', 'CODE(0x88560d8)') called at /usr/lib/perl5/vendor_perl/5.18.2/DBIx/Class/DeploymentHandler/DeployMethod/SQL/Translator.pm line 331
        DBIx::Class::DeploymentHandler::DeployMethod::SQL::Translator::txn_do('DBIx::Class::DeploymentHandler::DeployMethod::SQL::Translator...', 'CODE(0x8499308)') called at /usr/lib/perl5/vendor_perl/5.18.2/x86_64-linux-thread-multi/Moose/Meta/Method/Delegation.pm line 110
        DBIx::Class::DeploymentHandler::txn_do('DBIx::Class::DeploymentHandler=HASH(0x84b1470)', 'CODE(0x8499308)') called at /usr/lib/perl5/vendor_perl/5.18.2/DBIx/Class/DeploymentHandler/Dad.pm line 73
        DBIx::Class::DeploymentHandler::Dad::upgrade('DBIx::Class::DeploymentHandler=HASH(0x84b1470)') called at /usr/share/openqa/script/upgradedb line 137
DBIx::Class::Storage::TxnScopeGuard::DESTROY(): A DBIx::Class::Storage::TxnScopeGuard went out of scope without explicit commit or error. Rolling back. at /usr/lib/perl5/vendor_perl/5.18.2/DBIx/Class/DeploymentHandler/DeployMethod/SQL/Translator.pm line 321
DBIx::Class::Storage::TxnScopeGuard::DESTROY(): A DBIx::Class::Storage::TxnScopeGuard went out of scope without explicit commit or error. Rolling back. at /usr/lib/perl5/vendor_perl/5.18.2/DBIx/Class/DeploymentHandler/DeployMethod/SQL/Translator.pm line 321

**Comment by coolo** [comment]
Great stuff. I tested only with SQLite database from tortuga - the PG database proved to be too much for my laptop ;(

So what now?

**Comment by coolo** [comment]
the real problem is that it actually does 'soft_fails != 0 is NULL'. So my usage of raw SQL does not translate. I will switch to real raw SQL by means of UPDATE calls.

To make this reasonable, I'll drop MySQL - which we never used nor tested

**Comment by okurz** [comment]
@coolo maybe someone is using mysql and not telling us?

so there is "perl abstracted SQL" > "raw SQL" > "real raw SQL"?

**Comment by coolo** [comment]
if someone is using mysql and is not telling us, it can be brought back. But unless proven differently, I assume noone does.

**Comment by aaannz** [comment]
I'm little confused now, but we do consider soft-failed as passed, right? Because `OpenQA::Schema::Result::Jobs::calculate_result` wasn't updated so it now prints softfails as failed - preventing next run of chained tests. I wonder if scheduler doesn't need some update too for dependencies to work again.

**Comment by coolo** [comment]
yeah, are most likely several such cases. be it in openqa itself and external tools like the dashboard

**Comment by coolo** [comment]
updating calculate_result is asked too much right now - that's an art ;)

**Comment by mimi1vx** [comment]
this blocks qam-kernel tests  also:(

**Comment by coolo** [comment]
I can't see anything in that link that is related to this PR

**Comment by aaannz** [comment]
Should be fixed by https://github.com/os-autoinst/openQA/pull/801

File: lib/OpenQA/Schema/Result/Jobs.pm [hunk 2]
```diff
@@ -50,6 +50,7 @@ use constant FINAL_STATES     => (DONE,    CANCELLED);
 use constant {
     NONE               => 'none',
     PASSED             => 'passed',
+    SOFTFAILED         => 'softfailed',
```
Comment (AdamWill) [inline lib/OpenQA/Schema/Result/Jobs.pm]: See that comment a few lines earlier where it says "The state and results constants are duplicated in the Python client: if you change them or add any, please also update const.py." ? Well, I don't see a pull request for openQA-python-client :/ good thing I spotted this...

Comment (coolo) [inline lib/OpenQA/Schema/Result/Jobs.pm]: which proves me right when I say no one reads comments :)

Comment (okurz) [inline lib/OpenQA/Schema/Result/Jobs.pm]: +1, you need unit tests ;-)

**Pull request merged**